### PR TITLE
SNOW-913939: Align column name in local testing with live connection

### PR DESCRIPTION
**Pull request by sfc-gh-aling**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes SNOW-913939

handling double quotes in column name

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

File: src/snowflake/snowpark/mock/analyzer.py [hunk 1]
```diff
@@ -268,7 +273,7 @@ def analyze(
             sql = str(expr.value)
             if parse_local_name:
                 sql = sql.upper()
-            return sql
+            return f"'{sql}'" if quote_literal else sql
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/analyzer.py:276]: Why do we need this change?

File: src/snowflake/snowpark/_internal/analyzer/analyzer_utils.py [hunk 2]
```diff
@@ -1285,6 +1285,17 @@ def escape_quotes(unescaped: str) -> str:
     return unescaped.replace(DOUBLE_QUOTE, DOUBLE_QUOTE + DOUBLE_QUOTE)
 
 
+def uppercase_and_enquote_if_not_quoted(string):
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/_internal/analyzer/analyzer_utils.py:1288]: How is this function different than `quote_name`?

File: src/snowflake/snowpark/mock/connection.py [hunk 3]
```diff
@@ -355,7 +357,7 @@ def execute(
 
         res = execute_mock_plan(plan)
         if isinstance(res, TableEmulator):
-            columns = [*res.columns]
+            columns = [unquote_if_quoted(col_name) for col_name in res.columns]
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/connection.py:360]: Could you explain the reasoning behind this change? 🙂

File: src/snowflake/snowpark/mock/plan.py [hunk 4]
```diff
@@ -676,9 +685,10 @@ def describe(plan: MockExecutionPlan) -> List[Attribute]:
                 data_type = LongType()
             elif isinstance(data_type, FloatType):
                 data_type = DoubleType()
+
             ret.append(
                 Attribute(
-                    result[c].name,
+                    uppercase_and_enquote_if_not_quoted(result[c].name.strip()),
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/plan.py:691]: Looks like `uppercase_and_enquote_if_not_quoted` and `strip` is always used together. Maybe we should put `strip` in `uppercase_and_enquote_if_not_quoted`?

File: src/snowflake/snowpark/mock/plan.py [hunk 5]
```diff
@@ -708,7 +718,13 @@ def calculate_expression(
             # expr_id maps to the projected name, but input_data might still have the exp.name
             # dealing with the KeyError here, this happens in case df.union(df)
             # TODO: check SNOW-831880 for more context
-            return input_data[exp.name]
+            try:
+                # TODO: quoting is inconsistent during calculation, some columns are quoted while some are not
+                #  we need to decouple the logic between intermediate calculation and result representation,
+                #  check SNOW-942397 for more context
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/plan.py:724]: This sounds very error prone, should we aim to address this in PuPr instead?

File: tests/integ/test_column_names.py [hunk 6]
```diff
@@ -25,32 +25,49 @@
...
 def get_metadata_names(session, df):
+    if isinstance(session._conn, MockServerConnection):
+        return [col.name for col in session._conn.get_result_and_metadata(df._plan)[1]]
+
     description = session._conn._cursor.describe(df.queries["queries"][-1])
     return [quote_name(metadata.name) for metadata in description]
 
 
-def test_like(session):
-    df1 = session.sql("select 'v' as c")
```
Comment (sfc-gh-stan) [inline tests/integ/test_column_names.py:43]: I think we can just replace the original line with ` session.create_dataframe`.

Comment (sfc-gh-aling) [reply tests/integ/test_column_names.py:43]: you are right! updated

File: tests/integ/test_column_names.py [hunk 7]
```diff
@@ -25,32 +25,49 @@
...
     when,
 )
+from snowflake.snowpark.mock.connection import MockServerConnection
 from tests.utils import Utils
 
 pytestmark = pytest.mark.xfail(
     condition="config.getvalue('local_testing_mode')",
     raises=NotImplementedError,
-    strict=True,
+    strict=False,
```
[the remark this answers is not in the document]
Comment (sfc-gh-aling) [reply tests/integ/test_column_names.py]: updated

File: src/snowflake/snowpark/_internal/analyzer/analyzer_utils.py (hunk 2 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/_internal/analyzer/analyzer_utils.py]: no actually they are the same.. I removed this function and reused `quote_name`

File: src/snowflake/snowpark/mock/analyzer.py (hunk 1 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/analyzer.py:282]: I added some context in docstring and comment

File: src/snowflake/snowpark/mock/connection.py (hunk 3 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/connection.py:355]: I added some context in docstring and comment

File: src/snowflake/snowpark/mock/plan.py (hunk 4 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/plan.py]: I changed to use `quote_name`, so I will keep the `strip()` here

File: src/snowflake/snowpark/mock/plan.py (hunk 5 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/plan.py:723]: it is error prone.. the two logic are intertwined /facepalm/
I will see what I can do to improve

File: src/snowflake/snowpark/mock/analyzer.py (hunk 1 above)
Comment (sfc-gh-sfan) [reply src/snowflake/snowpark/mock/analyzer.py:282]: Do we need to worry if this is already quoted?

**Review by sfc-gh-sfan — APPROVED** [review]

File: src/snowflake/snowpark/mock/analyzer.py (hunk 1 above)
Comment (sfc-gh-stan) [reply src/snowflake/snowpark/mock/analyzer.py:282]: ^, I guess it happens when a string literal `'"hello"'`.

File: src/snowflake/snowpark/mock/analyzer.py [hunk 8]
```diff
@@ -172,6 +173,10 @@ def analyze(
             so that it will detect column name change.
             however, in the result calculation, we want to column name to be the output name, which is 'totB',
             so we set keep_alias to False in the execution.
+            quote_literal: if true, the literal string will be returned single quoted as '<literal>', column name
+            can be like e.g., <col> like '<literal>', <col> REGEXP '<literal>'.
+            quote_literal is set to True when `analyze` is called to describe the column name of a dataframe rather
+            than calculation, this is to align with the behavior of live connection describing a dataframe.
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/analyzer.py:179]: I found this a bit confusing, I thought `analyze` is always called to describe the resulting column name of an expression. `calculate_expression` is what actually computes.

File: src/snowflake/snowpark/mock/analyzer.py [hunk 9]
```diff
@@ -172,6 +173,10 @@ def analyze(
             so that it will detect column name change.
             however, in the result calculation, we want to column name to be the output name, which is 'totB',
             so we set keep_alias to False in the execution.
+            quote_literal: if true, the literal string will be returned single quoted as '<literal>', column name
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/analyzer.py:176]: Please correct me if I'm wrong: does `quote_literal` apply to any expressions that contain a literal value?

File: src/snowflake/snowpark/mock/plan.py [hunk 10]
```diff
@@ -167,8 +168,13 @@ def __init__(
 
 
 def execute_mock_plan(
-    plan: MockExecutionPlan, expr_to_alias: Optional[Dict[str, str]] = None
+    plan: MockExecutionPlan,
+    expr_to_alias: Optional[Dict[str, str]] = None,
 ) -> Union[TableEmulator, List[Row]]:
+    """
+
+    describe: if set to True, only get the column names without calculation
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/plan.py:176]: 🤔 `describe` is not a parameter here.

File: src/snowflake/snowpark/mock/plan.py (hunk 5 above)
Comment (sfc-gh-stan) [reply src/snowflake/snowpark/mock/plan.py:723]: Thanks for looking into it Adam! If you find it to be too much refactoring, I'm happy to keep it in post-PuPr.

**Review by sfc-gh-stan — CHANGES_REQUESTED** [review]

File: src/snowflake/snowpark/mock/plan.py (hunk 5 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/plan.py]: actually that's an issue in the aggregation where we set the column name, I found that we shall just add quotes for the intermediate results, such that quotes are aligned within our code.

thanks for pushing this

File: src/snowflake/snowpark/mock/analyzer.py (hunk 9 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/analyzer.py:176]: for now, `quote_literal=True` is only applied to the right child of `Like` and `RegExp` in my code change

File: src/snowflake/snowpark/mock/analyzer.py (hunk 8 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/analyzer.py:179]: yeah your understanding is correct.
I think this parameter needs a better name, probably something like  `quote_literal_in_column_name`.

so when we get the column name, single quote is added around the literal in the column name, e.g., in our test code "V%" is single quoted:
```python
def test_like(session):
    df1 = session.create_dataframe(["v"], schema=["c"])
    df2 = df1.select(df1["c"].like(lit("v%")))

    assert (
        df2._output[0].name
        == df2.columns[0]
        == get_metadata_names(session, df2)[0]
        == '"""C"" LIKE \'V%\'"'
    )
```

but for calculation, we are just comparing the string "%V", no single quotes.

Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/analyzer.py]: updated to use `calculate_expression` to literal computation and single quote literal all the time in analyze

File: src/snowflake/snowpark/mock/analyzer.py (hunk 9 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/analyzer.py]: updated to use `calculate_expression` to literal computation and single quote literal all the time in analyze

File: src/snowflake/snowpark/mock/plan.py (hunk 5 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/plan.py]: updated to use `calculate_expression` to literal computation and single quote literal all the time in analyze

File: src/snowflake/snowpark/mock/analyzer.py (hunk 1 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/analyzer.py]: updated to use `calculate_expression` to literal computation and single quote literal all the time in analyze.

**Pull request merged**